### ISSUE TITLE
missed optimization in scalar multiplication of array that is never used again

### ISSUE THREAD

**Issue by mrvn** [llvm/llvm-project#56291]
I have found a missed optimization in all versions of clang from 9.0 to 14.0.0 with -O2 and -O3. I have distilled the problem down to this test case:

```
constexpr int SIZE = 10000000;

static void mult(int* a, int b) {
    for (int i = 0; i < SIZE; i++) a[i] *= b;
}

int main() {
    int* a = new int[SIZE]{};
    mult(a, 5);
    delete[] a;
}

```

The array `a` is never used again after the call to `mult()` and `mult()` has no side effect apart from modifying `a`. The function call can be removed, which then allows the `new` and `delete` to be merged and removed.

Expected result: the equivalent of `int main() { }`
Actual result: `mult()` gets inlined and unrolled into a hugely complex SIMD operation.

PS: gcc manages to remove the function call.

**Comment by jrtc27** [comment]
The function call _is_ removed, and manually inlining doesn't change things. The problem is just your array is too big for Clang to keep track of; once it reaches 60 elements it doesn't reason about the data dependencies. https://godbolt.org/z/K4j3634qP is a simplified test case showing this.

**Comment by mrvn** [comment]
Not sure about the array size alone being the culprit because with the same array this code
```
for (int i = 0; i < 10000000; ++i) a[i] = rand();
```
gets optimized into just calling `rand()` 10000000 times. But here the array is only written to and not read and written like in the missed optimization.

**Comment by mrvn** [comment]
Indeed. Only the `mult` is the problem. I just need the allocation to have the array in the test case.

**Comment by jrtc27** [comment]
> The problem here is that we don't infer that `mult` is willreturn. If we did, then InstCombine would drop the allocation.

No it's not, mult was inlined so its attributes are totally irrelevant. The problem is whether the loop is short enough that LoopFullUnrollPass unrolls it fully, allowing GVN to figure out all the dead loads/stores, or not. Note also the threshold is different between -O2 and -O3 (UnrollThresholdDefault/unroll-threshold-default, 150 by default, and UnrollThresholdAggressive/unroll-threshold-aggressive, 300 by default).
